modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/SidebarPanel.js

```diff
@@ -12,30 +12,100 @@
  * details.
  */
 
+import ClayAlert from '@clayui/alert';
 import ClayLoadingIndicator from '@clayui/loading-indicator';
-import React, {useImperativeHandle, useState} from 'react';
+import {fetch} from 'frontend-js-web';
+import React, {
+	useCallback,
+	useEffect,
+	useImperativeHandle,
+	useRef,
+	useState,
+} from 'react';
 
 import Sidebar from './components/Sidebar';
 
-const SidebarPanel = React.forwardRef(({onClose}, ref) => {
-	const [isOpen, setIsOpen] = useState(true);
+const SidebarPanel = React.forwardRef(
+	({fetchURL, onClose, viewComponent: View}, ref) => {
+		const CurrentView = useRef(View);
 
-	useImperativeHandle(ref, () => ({
-		close: () => setIsOpen(false),
-		open: () => setIsOpen(true),
-	}));
+		const [error, setError] = useState();
+		const [isLoading, setIsLoading] = useState();
+		const [isOpen, setIsOpen] = useState(true);
+		const [resourceData, setResourceData] = useState();
 
-	const onCloseHandle = () => (onClose ? onClose() : setIsOpen(false));
+		const getData = useCallback((fetchURL) => {
+			setIsLoading(true);
+			setResourceData(null);
 
-	return (
-		<Sidebar onClose={onCloseHandle} open={isOpen}>
-			<Sidebar.Header />
+			fetch(fetchURL, {
+				method: 'GET',
+			})
+				.then((response) => response.json())
+				.then((data) => {
+					setData(data, data?.error);
+				})
+				.catch(() => {
+					setData(
+						null,
+						Liferay.Language.get('an-unexpected-error-occurred')
+					);
+				});
+		}, []);
 
-			<Sidebar.Body>
-				<ClayLoadingIndicator />
-			</Sidebar.Body>
-		</Sidebar>
-	);
-});
+		const onCloseHandle = () => (onClose ? onClose() : setIsOpen(false));
+
+		const setData = (data, error) => {
+
+			// Force 300 ms of waiting to render the response so loading
+			// looks more natural.
+
+			setTimeout(() => {
+				setIsLoading(false);
+				setError(error);
+				setResourceData(data);
+			}, 300);
```

Comment:
@carloslancha: I'm late to the party on this one, but this is another example of a place where we should be using `useTimeout()` (safe: 🦄) instead of `setTimeout()` (dangerous: 💀) (like in this [follow-up](https://github.com/liferay-frontend/liferay-portal/pull/154)).

Also, whenever you see a bunch of `setSomething()` functions occurring in a batch like this, it is a cue that the whole thing might be better off modeled using `useReducer()`. That is, when there are data dependencies between the various "atoms" of state, it can make sense to just model it as a single complex "molecule" of state. This way, you ensure that incoherent combinations of state are impossible, because you have a single place (the reducer) that determines all possible combinations of state.

For example, you don't have to worry about what happens if `error` *and* `resourceData` are set at the same time etc.

Now, not saying that you should race and change this to use a reducer, but you should at least look over it and go through the mental exercise of looking at what it would be like if you did; depending on what that looks like, you either change it or not.

Reply:
Changing setTimeout here to, thanks for pointing it out! Will be on https://github.com/liferay-frontend/liferay-portal/pull/156

Will take a look and think about the other stuff 😄